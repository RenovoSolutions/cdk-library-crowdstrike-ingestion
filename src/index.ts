export * from './bucket';
export * from './logsubscription';