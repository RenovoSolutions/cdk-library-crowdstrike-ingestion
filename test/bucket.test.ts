import {
  Duration,
  Stack,
  aws_iam as iam,
} from 'aws-cdk-lib';
import {
  Annotations,
  Match,
  Template,
} from 'aws-cdk-lib/assertions';
import { CrowdStrikeBucket } from '../src/bucket';

describe('CrowdStrikeBucket', () => {
  let stack: Stack;

  beforeEach(() => {
    stack = new Stack();
  });

  test('creates bucket with default properties', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify bucket creation with expected properties
    template.hasResourceProperties('AWS::S3::Bucket', {
      BucketName: 'test-crowdstrike-bucket',
      BucketEncryption: {
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: 'AES256',
            },
          },
        ],
      },
      PublicAccessBlockConfiguration: {
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      },
      VersioningConfiguration: {
        Status: 'Enabled',
      },
    });

    // Verify SQS queue creation
    template.hasResourceProperties('AWS::SQS::Queue', {
      QueueName: {
        'Fn::Join': [
          '',
          [
            {
              Ref: Match.stringLikeRegexp('TestBucket'),
            },
            '-queue',
          ],
        ],
      },
    });

    // Verify DLQ creation
    template.hasResourceProperties('AWS::SQS::Queue', {
      QueueName: {
        'Fn::Join': [
          '',
          [
            {
              Ref: Match.stringLikeRegexp('TestBucket'),
            },
            '-dlq',
          ],
        ],
      },
    });

    // Verify IAM role creation
    template.hasResourceProperties('AWS::IAM::Role', {
      RoleName: {
        'Fn::Join': [
          '',
          [
            {
              Ref: Match.stringLikeRegexp('TestBucket'),
            },
            '-role',
          ],
        ],
      },
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Condition: {
              StringEquals: {
                'sts:ExternalId': {
                  Ref: Match.stringLikeRegexp('TestBucketCrowdStrikeExternalIdParam'),
                },
              },
            },
            Effect: 'Allow',
            Principal: {
              AWS: {
                Ref: Match.stringLikeRegexp('TestBucketCrowdStrikeRoleParam'),
              },
            },
          },
        ],
      },
    });

    // Verify bucket notification configuration
    template.hasResourceProperties('Custom::S3BucketNotifications', {
      NotificationConfiguration: {
        QueueConfigurations: [
          {
            Events: [
              's3:ObjectCreated:*',
            ],
          },
        ],
      },
    });

    // Stack outputs
    template.hasOutput('*', {
      Value: {
        Ref: Match.stringLikeRegexp('TestBucket'),
      },
      Description: 'The Name of the S3 bucket for CrowdStrike ingestion',
    });

    template.hasOutput('*', {
      Value: {
        'Fn::GetAtt': [
          Match.stringLikeRegexp('TestBucket'),
          'Arn',
        ],
      },
      Description: 'The ARN of the S3 bucket for CrowdStrike ingestion',
    });

    template.hasOutput('*', {
      Value: {
        'Fn::GetAtt': [
          Match.stringLikeRegexp('TestBucketQueue'),
          'QueueName',
        ],
      },
      Description: 'The Name of the SQS queue for CrowdStrike ingestion',
    });

    template.hasOutput('*', {
      Value: {
        Ref: Match.stringLikeRegexp('TestBucketRole'),
      },
      Description: 'The Name of the IAM role for CrowdStrike ingestion',
    });

    // Template snapshot
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('creates bucket with KMS key', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
      createKmsKey: true,
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify KMS key creation
    template.hasResourceProperties('AWS::KMS::Key', {
      Description: {
        'Fn::Join': [
          '',
          [
            'KMS Key for CrowdStrike ingestion bucket ',
            {
              Ref: Match.stringLikeRegexp('TestBucket'),
            },
          ],
        ],
      },
      EnableKeyRotation: false,
      MultiRegion: true,
    });

    // Verify KMS alias
    template.hasResourceProperties('AWS::KMS::Alias', {
      AliasName: {
        'Fn::Join': [
          '',
          [
            'alias/',
            {
              Ref: Match.stringLikeRegexp('TestBucket'),
            },
          ],
        ],
      },
    });

    // Additional Stack output
    template.hasOutput('*', {
      Value: {
        'Fn::GetAtt': [
          Match.stringLikeRegexp('TestBucketKey'),
          'Arn',
        ],
      },
      Description: 'The ARN of the KMS key for CrowdStrike ingestion',
    });
  });

  test('creates bucket with custom queue name and properties', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
      queueProps: {
        queueName: 'custom-queue-name',
        visibilityTimeout: Duration.seconds(300),
      },
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify SQS queue with custom name and properties
    template.hasResourceProperties('AWS::SQS::Queue', {
      QueueName: 'custom-queue-name',
      VisibilityTimeout: 300,
    });
  });

  test('creates bucket with custom role name and properties', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      roleProps: {
        roleName: 'custom-role-name',
        description: 'Custom role description',
        assumedBy: new iam.PrincipalWithConditions(new iam.ArnPrincipal('arn:aws:iam::123456789012:role/CrowdStrikeRole'), {
          StringEquals: {
            'sts:ExternalId': 'externalId123',
          },
        }),
      },
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify IAM role with custom name and properties
    template.hasResourceProperties('AWS::IAM::Role', {
      RoleName: 'custom-role-name',
      Description: 'Custom role description',
    });
  });

  test('creates bucket with custom KMS key name and properties', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
      createKmsKey: true,
      keyProps: {
        alias: 'custom-key-alias',
        description: 'Custom key description',
        enableKeyRotation: true,
      },
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify KMS key with custom properties
    template.hasResourceProperties('AWS::KMS::Key', {
      Description: 'Custom key description',
      EnableKeyRotation: true,
    });

    template.hasResourceProperties('AWS::KMS::Alias', {
      AliasName: 'alias/custom-key-alias',
    });
  });

  test('creates bucket with organization access', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
      orgIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/orgId/path',
      createKmsKey: true,
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify bucket policy allows organization access
    template.hasResourceProperties('AWS::S3::BucketPolicy', {
      PolicyDocument: {
        Statement: [
          {
            Action: 's3:*',
            Condition: {
              Bool: {
                'aws:SecureTransport': 'false',
              },
            },
            Effect: 'Deny',
            Principal: {
              AWS: '*',
            },
            Resource: [
              {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('TestBucket'),
                  'Arn',
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('TestBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
            ],
          },
          {
            Action: [
              's3:DeleteObject*',
              's3:PutObject',
              's3:PutObjectLegalHold',
              's3:PutObjectRetention',
              's3:PutObjectTagging',
              's3:PutObjectVersionTagging',
              's3:Abort*',
            ],
            Effect: 'Allow',
            Principal: {
              AWS: '*',
            },
            Resource: [
              {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('TestBucket'),
                  'Arn',
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('TestBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
            ],
            Condition: {
              StringEquals: {
                'aws:PrincipalOrgID': {
                  Ref: Match.stringLikeRegexp('TestBucketOrgIdParamParameter'),
                },
              },
            },
          },
        ],
      },
    });

    // Verify KMS key policy allows organization access
    template.hasResourceProperties('AWS::KMS::Key', {
      KeyPolicy: {
        Statement: [
          {
            Action: 'kms:*',
            Effect: 'Allow',
            Principal: {
              AWS: {
                'Fn::Join': [
                  '',
                  [
                    'arn:',
                    {
                      Ref: 'AWS::Partition',
                    },
                    ':iam::',
                    {
                      Ref: 'AWS::AccountId',
                    },
                    ':root',
                  ],
                ],
              },
            },
            Resource: '*',
          },
          {
            Action: [
              'kms:Decrypt',
              'kms:Encrypt',
              'kms:ReEncrypt*',
              'kms:GenerateDataKey*',
            ],
            Effect: 'Allow',
            Principal: {
              AWS: '*',
            },
            Resource: '*',
            Condition: {
              StringEquals: {
                'aws:PrincipalOrgID': {
                  Ref: Match.stringLikeRegexp('TestBucketOrgIdParamParameter'),
                },
              },
            },
          },
        ],
      },
    });
  });

  test('creates bucket with logging bucket source', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      loggingBucketSourceName: 'source-logging-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify bucket policy allows logging from source bucket
    template.hasResourceProperties('AWS::S3::BucketPolicy', {
      PolicyDocument: {
        Statement: [
          {
            Action: 's3:*',
            Effect: 'Deny',
            Principal: {
              AWS: '*',
            },
            Resource: [
              {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('TestBucket'),
                  'Arn',
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('TestBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
            ],
            Condition: {
              Bool: {
                'aws:SecureTransport': 'false',
              },
            },
          },
          {
            Action: 's3:PutObject',
            Effect: 'Allow',
            Principal: {
              Service: 'logging.s3.amazonaws.com',
            },
            Resource: {
              'Fn::Join': [
                '',
                [
                  {
                    'Fn::GetAtt': [
                      Match.stringLikeRegexp('TestBucket'),
                      'Arn',
                    ],
                  },
                  '/*',
                ],
              ],
            },
            Condition: {
              ArnLike: {
                'aws:SourceArn': 'arn:aws:s3:::source-logging-bucket',
              },
              StringEquals: {
                'aws:SourceAccount': {
                  Ref: 'AWS::AccountId',
                },
              },
            },
          },
        ],
      },
    });
  });

  test('bucket has correct removal policy', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify bucket has RETAIN removal policy
    template.hasResource('AWS::S3::Bucket', {
      DeletionPolicy: 'RetainExceptOnCreate',
      UpdateReplacePolicy: 'Retain',
    });
  });

  test('queue and role have correct permissions', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
      crowdStrikeRoleParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/role/path',
      crowdStrikeExternalIdParameterArn: 'arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path',
    });

    // THEN
    const template = Template.fromStack(stack);

    // Verify role has permissions to read from bucket and consume messages from SQS queue
    template.hasResourceProperties('AWS::IAM::Policy', {
      PolicyDocument: {
        Statement: [
          {
            Action: [
              's3:GetObject*',
              's3:GetBucket*',
              's3:List*',
            ],
            Effect: 'Allow',
            Resource: [
              {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('TestBucket'),
                  'Arn',
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('TestBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
            ],
          },
          {
            Action: [
              'sqs:ReceiveMessage',
              'sqs:ChangeMessageVisibility',
              'sqs:GetQueueUrl',
              'sqs:DeleteMessage',
              'sqs:GetQueueAttributes',
            ],
            Effect: 'Allow',
            Resource: {
              'Fn::GetAtt': [
                Match.stringLikeRegexp('TestBucketQueue'),
                'Arn',
              ],
            },
          },
        ],
      },
    });
  });

  test('annotates error if required parameters are missing', () => {
    // WHEN
    new CrowdStrikeBucket(stack, 'TestBucket', {
      bucketName: 'test-crowdstrike-bucket',
    });

    // THEN
    Annotations.fromStack(stack).hasError(
      '/Default/TestBucket',
      'Either roleProps or both crowdStrikeRoleParameterArn and crowdStrikeExternalIdParameterArn must be provided.',
    );
  });
});