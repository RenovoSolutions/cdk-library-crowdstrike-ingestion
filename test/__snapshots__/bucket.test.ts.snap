// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrowdStrikeBucket creates bucket with default properties 1`] = `
{
  "Outputs": {
    "TestBucketBucketArnFAE9782D": {
      "Description": "The ARN of the S3 bucket for CrowdStrike ingestion",
      "Value": {
        "Fn::GetAtt": [
          "TestBucket560B80BC",
          "Arn",
        ],
      },
    },
    "TestBucketBucketName6CABE3C3": {
      "Description": "The Name of the S3 bucket for CrowdStrike ingestion",
      "Value": {
        "Ref": "TestBucket560B80BC",
      },
    },
    "TestBucketQueueNameD0D0DC9E": {
      "Description": "The Name of the SQS queue for CrowdStrike ingestion",
      "Value": {
        "Fn::GetAtt": [
          "TestBucketQueue8E1A8921",
          "QueueName",
        ],
      },
    },
    "TestBucketRoleName8AF73023": {
      "Description": "The Name of the IAM role for CrowdStrike ingestion",
      "Value": {
        "Ref": "TestBucketRoleCF6B6A2F",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "TestBucketCrowdStrikeExternalIdParamParameterB69C278D": {
      "Default": "arn:aws:ssm:us-east-1:123456789012:parameter/custom/externalId/path",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": {
      "DependsOn": [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "import boto3  # type: ignore
import json
import logging
import urllib.request

s3 = boto3.client("s3")

EVENTBRIDGE_CONFIGURATION = 'EventBridgeConfiguration'
CONFIGURATION_TYPES = ["TopicConfigurations", "QueueConfigurations", "LambdaFunctionConfigurations"]

def handler(event: dict, context):
  response_status = "SUCCESS"
  error_message = ""
  try:
    props = event["ResourceProperties"]
    notification_configuration = props["NotificationConfiguration"]
    managed = props.get('Managed', 'true').lower() == 'true'
    skipDestinationValidation = props.get('SkipDestinationValidation', 'false').lower() == 'true'
    stack_id = event['StackId']
    old = event.get("OldResourceProperties", {}).get("NotificationConfiguration", {})
    if managed:
      config = handle_managed(event["RequestType"], notification_configuration)
    else:
      config = handle_unmanaged(props["BucketName"], stack_id, event["RequestType"], notification_configuration, old)
    s3.put_bucket_notification_configuration(Bucket=props["BucketName"], NotificationConfiguration=config, SkipDestinationValidation=skipDestinationValidation)
  except Exception as e:
    logging.exception("Failed to put bucket notification configuration")
    response_status = "FAILED"
    error_message = f"Error: {str(e)}. "
  finally:
    submit_response(event, context, response_status, error_message)

def handle_managed(request_type, notification_configuration):
  if request_type == 'Delete':
    return {}
  return notification_configuration

def handle_unmanaged(bucket, stack_id, request_type, notification_configuration, old):
  def get_id(n):
    n['Id'] = ''
    sorted_notifications = sort_filter_rules(n)
    strToHash=json.dumps(sorted_notifications, sort_keys=True).replace('"Name": "prefix"', '"Name": "Prefix"').replace('"Name": "suffix"', '"Name": "Suffix"')
    return f"{stack_id}-{hash(strToHash)}"
  def with_id(n):
    n['Id'] = get_id(n)
    return n

  external_notifications = {}
  existing_notifications = s3.get_bucket_notification_configuration(Bucket=bucket)
  for t in CONFIGURATION_TYPES:
    if request_type == 'Update':
        old_incoming_ids = [get_id(n) for n in old.get(t, [])]
        external_notifications[t] = [n for n in existing_notifications.get(t, []) if not get_id(n) in old_incoming_ids]      
    elif request_type == 'Delete':
        external_notifications[t] = [n for n in existing_notifications.get(t, []) if not n['Id'].startswith(f"{stack_id}-")]
    elif request_type == 'Create':
        external_notifications[t] = [n for n in existing_notifications.get(t, [])]
  if EVENTBRIDGE_CONFIGURATION in existing_notifications:
    external_notifications[EVENTBRIDGE_CONFIGURATION] = existing_notifications[EVENTBRIDGE_CONFIGURATION]

  if request_type == 'Delete':
    return external_notifications

  notifications = {}
  for t in CONFIGURATION_TYPES:
    external = external_notifications.get(t, [])
    incoming = [with_id(n) for n in notification_configuration.get(t, [])]
    notifications[t] = external + incoming

  if EVENTBRIDGE_CONFIGURATION in notification_configuration:
    notifications[EVENTBRIDGE_CONFIGURATION] = notification_configuration[EVENTBRIDGE_CONFIGURATION]
  elif EVENTBRIDGE_CONFIGURATION in external_notifications:
    notifications[EVENTBRIDGE_CONFIGURATION] = external_notifications[EVENTBRIDGE_CONFIGURATION]

  return notifications

def submit_response(event: dict, context, response_status: str, error_message: str):
  response_body = json.dumps(
    {
      "Status": response_status,
      "Reason": f"{error_message}See the details in CloudWatch Log Stream: {context.log_stream_name}",
      "PhysicalResourceId": event.get("PhysicalResourceId") or event["LogicalResourceId"],
      "StackId": event["StackId"],
      "RequestId": event["RequestId"],
      "LogicalResourceId": event["LogicalResourceId"],
      "NoEcho": False,
    }
  ).encode("utf-8")
  headers = {"content-type": "", "content-length": str(len(response_body))}
  try:
    req = urllib.request.Request(url=event["ResponseURL"], headers=headers, data=response_body, method="PUT")
    with urllib.request.urlopen(req) as response:
      print(response.read().decode("utf-8"))
    print("Status code: " + response.reason)
  except Exception as e:
      print("send(..) failed executing request.urlopen(..): " + str(e))

def sort_filter_rules(json_obj):
  if not isinstance(json_obj, dict):
      return json_obj
  for key, value in json_obj.items():
      if isinstance(value, dict):
          json_obj[key] = sort_filter_rules(value)
      elif isinstance(value, list):
          json_obj[key] = [sort_filter_rules(item) for item in value]
  if "Filter" in json_obj and "Key" in json_obj["Filter"] and "FilterRules" in json_obj["Filter"]["Key"]:
      filter_rules = json_obj["Filter"]["Key"]["FilterRules"]
      sorted_filter_rules = sorted(filter_rules, key=lambda x: x["Name"])
      json_obj["Filter"]["Key"]["FilterRules"] = sorted_filter_rules
  return json_obj",
        },
        "Description": "AWS CloudFormation handler for "Custom::S3BucketNotifications" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": [
          {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestBucket560B80BC": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "test-crowdstrike-bucket",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestBucketDLQ4922D481": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "TestBucket560B80BC",
              },
              "-dlq",
            ],
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "TestBucketDLQPolicyA30AE17E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "TestBucketDLQ4922D481",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "TestBucketDLQ4922D481",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "TestBucketNotificationsE54C04C2": {
      "DependsOn": [
        "TestBucketPolicyBA12ED38",
        "TestBucketQueuePolicy37772A6A",
        "TestBucketQueue8E1A8921",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "TestBucket560B80BC",
        },
        "Managed": true,
        "NotificationConfiguration": {
          "QueueConfigurations": [
            {
              "Events": [
                "s3:ObjectCreated:*",
              ],
              "QueueArn": {
                "Fn::GetAtt": [
                  "TestBucketQueue8E1A8921",
                  "Arn",
                ],
              },
            },
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
        "SkipDestinationValidation": false,
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "TestBucketPolicyBA12ED38": {
      "Properties": {
        "Bucket": {
          "Ref": "TestBucket560B80BC",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestBucket560B80BC",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestBucket560B80BC",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestBucketQueue8E1A8921": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "TestBucket560B80BC",
              },
              "-queue",
            ],
          ],
        },
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "TestBucketDLQ4922D481",
              "Arn",
            ],
          },
          "maxReceiveCount": 5,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "TestBucketQueuePolicy37772A6A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "TestBucketQueue8E1A8921",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Condition": {
                "ArnLike": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "TestBucket560B80BC",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "s3.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "TestBucketQueue8E1A8921",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "TestBucketQueue8E1A8921",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "TestBucketRoleCF6B6A2F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "StringEquals": {
                  "sts:ExternalId": {
                    "Ref": "TestBucketCrowdStrikeExternalIdParamParameterB69C278D",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "arn:aws:iam::123456789012:role/crowdstrike-role-here",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "TestBucket560B80BC",
              },
              "-role",
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestBucketRoleDefaultPolicy3121C833": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestBucket560B80BC",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestBucket560B80BC",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TestBucketQueue8E1A8921",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestBucketRoleDefaultPolicy3121C833",
        "Roles": [
          {
            "Ref": "TestBucketRoleCF6B6A2F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
